Fixed error handling in `RequestExtensions`

holterbades opened the pull request:
Response error handling was unintentionally removed in a previous commit. Added it back to the `ExecuteTaskCoreAsync` method. Fixed `NullReferenceException`-bug in the `HandleIfErrorResponse` method - `throw response.Exception;` does not make sense if the exception is null.

Fixes #1041 , #1020 , and probably #1027

holterbades commented:
To the reviewer: I also noticed a minor codesmell: In line 110 and 113 there are null-checks to the `path` variable, which are unnecessary, because the assignment via `Split('/')` doesn't return `null`.

ebozduman commented:
I'll start reviewing this in the morning.

holterbades commented:
Maybe i gonna add some tests so that this error does not happen again.
Is there a special reason why in `FunctionalTest` only Streams are used as data source (`GenerateStreamFromSeed`)?
I would like to add a consistency check with an hash value or a full content check to the test case `Minio.Functional.Tests.FunctionalTest.GetObject_Test2` so that there is a validation that the upload matches the download.

ebozduman commented:
Sounds great @holterbades !

ebozduman commented:
> To the reviewer: I also noticed a minor codesmell: In line 110 and 113 there are null-checks to the `path` variable, which are unnecessary, because the assignment via `Split('/')` doesn't return `null`.

Those are not null checks. 
They are there to see if there is 1 or more than 1 elements in the `path`.
So, 1 means the command deals with a bucket, and more than 1 element in the `path` means the command deals with an object.

holterbades commented:
> Those are not null checks.
They are there to see if there is 1 or more than 1 elements in the path.

Sure there are null-checks; I mean the `?` in the `path?.ToList()`. If you have `var path = someUri.Split('/');` in line 104, `path` can't be null because `Split(...)` does not return `null`. The least `.Split()` can return is a `string[1] { "" }` if the input is `""`.

Also it would be more efficient to access the element count of an array via `path.Length` and not convert it to a list twice. 

I changed it accordingly.

holterbades commented:
I would hand in changes to the tests in a separate PullRequest from my private account some time later.

holterbades commented:
I don't get it why the assertion is now failing but did not fail the last executions of this testrun before the error handling bug . But in my opinion this testcase is wrong: The server responds with an error, which is expected in this test case, but the error message which is returned is checked if it is exactly the expected message. In my opinion the test case should only check if the error message starts with the expected text.

Furthermore there was another test bug which was introduced in the same commit mentioned in #1041 and was not discovered because of the removed error handling.

holterbades commented:
hmmmh ... The test is now failing at a new location. Locally this testcase is working just fine. Maybe a race condition in the test itself? I really don't think this has something to do with my fix.

holterbades commented:
@ebozduman Hey, I would need some help with that PR. I don't know what is going on with this tests.

ebozduman commented:
@holterbades 

There is another PR waiting for to be reviewed, PR#1043, and I think it can help resolve some of these test failure issues.
Let's wait for it to be merged, then rebase your/this PR and let's see how it goes with this PR.
You are more then welcome to check it out and review it, if you have time.

holterbades commented:
@ebozduman 
Since I suspect that this #1043 will take even longer, I would strongly suggest [unlisting the currently broken nuget package from nuget.org](https://learn.microsoft.com/en-us/nuget/nuget-org/policies/deleting-packages), so that no user can install this broken version.

ebozduman commented:
I just ran `dotnet regitlint` on your branch for the format issue(s).
I think the build test will run smoothly and all finish successfully.
Then I can merge the PR.

ebozduman reviewed (approved):
LGTM
